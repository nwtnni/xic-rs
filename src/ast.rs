use crate::span::Span;
use crate::symbol;

/// Represents a Xi interface file.
#[derive(Clone, Debug)]
pub struct Interface {
    pub sigs: Vec<Sig>,
}

/// Represents a Xi source file.
#[derive(Clone, Debug)]
pub struct Program {
    pub uses: Vec<Use>,
    pub funs: Vec<Fun>,
}

/// Represents a use statement for importing interfaces.
#[derive(Clone, Debug)]
pub struct Use {
    pub name: symbol::Symbol,
    pub span: Span,
}

/// Represents a function signature (i.e. without implementation).
#[derive(Clone, Debug)]
pub struct Sig {
    pub name: symbol::Symbol,
    pub args: Vec<Dec>,
    pub rets: Vec<Typ>,
    pub span: Span,
}

/// Represents a function definition (i.e. with implementation).
#[derive(Clone, Debug)]
pub struct Fun {
    pub name: symbol::Symbol,
    pub args: Vec<Dec>,
    pub rets: Vec<Typ>,
    pub body: Stm,
    pub span: Span,
}

/// Represents a primitive type.
#[derive(Clone, Debug)]
pub enum Typ {
    Bool(Span),
    Int(Span),
    Arr(Box<Typ>, Option<Exp>, Span),
}

impl PartialEq for Typ {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
        | (Typ::Bool(_), Typ::Bool(_))
        | (Typ::Int(_), Typ::Int(_)) => true,
        | (Typ::Arr(lhs, _, _), Typ::Arr(rhs, _, _)) => lhs == rhs,
        | _ => false,
        }
    }
}

impl Eq for Typ {}

/// Represents a binary operator.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Bin {
    Mul,
    Hul,
    Div,
    Mod,
    Add,
    Sub,
    Lt,
    Le,
    Ge,
    Gt,
    Eq,
    Ne,
    And,
    Or,
}

/// Represents a unary operator.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Uno {
    Neg,
    Not,
}

/// Represents an expression (i.e. a term that can be evaluated).
#[derive(Clone, Debug)]
pub enum Exp {
    /// Boolean literal
    Bool(bool, Span),

    /// Char literal
    Chr(char, Span),

    /// String literal
    Str(String, Span),

    /// Integer literal
    Int(i64, Span),

    /// Variable
    Var(symbol::Symbol, Span),

    /// Array literal
    Arr(Vec<Exp>, Span),

    /// Binary operation
    Bin(Bin, Box<Exp>, Box<Exp>, Span),

    /// Unary operation
    Uno(Uno, Box<Exp>, Span),

    /// Array index
    Idx(Box<Exp>, Box<Exp>, Span),

    /// Function call
    Call(Call),
}

/// Represents a variable declaration.
#[derive(Clone, Debug)]
pub struct Dec {
    pub name: symbol::Symbol,
    pub typ: Typ,
    pub span: Span,
}

/// Represents a function call.
#[derive(Clone, Debug)]
pub struct Call {
    pub name: symbol::Symbol,
    pub args: Vec<Exp>,
    pub span: Span,
}

/// Represents an imperative statement.
#[derive(Clone, Debug)]
pub enum Stm {
    /// Assignment
    Ass(Exp, Exp, Span),

    /// Procedure call
    Call(Call),

    /// Initialization
    Init(Vec<Option<Dec>>, Exp, Span),

    /// Variable declaration
    Dec(Dec, Span),
    
    /// Return statement
    Ret(Vec<Exp>, Span),

    /// Statement block
    Seq(Vec<Stm>, Span),

    /// If-else block
    If(Exp, Box<Stm>, Option<Box<Stm>>, Span),

    /// While block
    While(Exp, Box<Stm>, Span),
}
