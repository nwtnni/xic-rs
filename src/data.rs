pub mod token;
pub mod ast;
pub mod typ;
pub mod operand;
