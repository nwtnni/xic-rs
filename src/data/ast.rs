use std::cell::Cell;
use std::fmt;

use crate::data::sexp::Serialize as _;
use crate::data::span::Span;
use crate::data::symbol::Symbol;

/// Represents a Xi interface file.
#[derive(Clone, Debug)]
pub struct Interface {
    pub uses: Vec<Use>,
    pub items: Vec<ItemSignature>,
}

impl fmt::Display for Interface {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a Xi source file.
#[derive(Clone, Debug)]
pub struct Program {
    pub uses: Vec<Use>,
    pub items: Vec<Item>,
}

impl fmt::Display for Program {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a use statement for importing interfaces.
#[derive(Clone, Debug)]
pub struct Use {
    pub name: Symbol,
    pub span: Span,
}

impl fmt::Display for Use {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum ItemSignature {
    Class(ClassSignature),
    Function(FunctionSignature),
}

impl fmt::Display for ItemSignature {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum Item {
    Global(Global),
    Class(Class),
    Function(Function),
}

impl fmt::Display for Item {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum Global {
    Declaration(Declaration),
    Initialization(Initialization),
}

impl fmt::Display for Global {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct Initialization {
    pub declarations: Vec<Option<SingleDeclaration>>,
    pub expression: Expression,
    pub span: Span,
}

impl fmt::Display for Initialization {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct ClassSignature {
    pub name: Symbol,
    pub extends: Option<Symbol>,
    pub methods: Vec<FunctionSignature>,
    pub span: Span,
}

impl fmt::Display for ClassSignature {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct Class {
    pub name: Symbol,
    pub extends: Option<Symbol>,
    pub items: Vec<ClassItem>,
    pub span: Span,
}

impl fmt::Display for Class {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum ClassItem {
    Field(Declaration),
    Method(Function),
}

impl fmt::Display for ClassItem {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

pub trait Callable {
    fn name(&self) -> Symbol;
    fn parameters(&self) -> &[SingleDeclaration];
    fn returns(&self) -> &[Type];
}

macro_rules! impl_callable {
    ($type:ty) => {
        impl Callable for $type {
            fn name(&self) -> Symbol {
                self.name
            }

            fn parameters(&self) -> &[SingleDeclaration] {
                &self.parameters
            }

            fn returns(&self) -> &[Type] {
                &self.returns
            }
        }
    };
}

/// Represents a function signature (i.e. without implementation).
#[derive(Clone, Debug)]
pub struct FunctionSignature {
    pub name: Symbol,
    pub parameters: Vec<SingleDeclaration>,
    pub returns: Vec<Type>,
    pub span: Span,
}

impl_callable!(FunctionSignature);

impl fmt::Display for FunctionSignature {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a function definition (i.e. with implementation).
#[derive(Clone, Debug)]
pub struct Function {
    pub name: Symbol,
    pub parameters: Vec<SingleDeclaration>,
    pub returns: Vec<Type>,
    pub statements: Statement,
    pub span: Span,
}

impl_callable!(Function);

impl fmt::Display for Function {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a primitive type.
#[derive(Clone, Debug)]
pub enum Type {
    Bool(Span),
    Int(Span),
    Class(Symbol, Span),
    Array(Box<Type>, Option<Expression>, Span),
}

impl Type {
    pub fn has_length(&self) -> bool {
        match self {
            Type::Bool(_) | Type::Int(_) | Type::Class(_, _) => false,
            Type::Array(r#type, length, _) => length.is_some() || r#type.has_length(),
        }
    }

    pub fn span(&self) -> Span {
        match self {
            Type::Bool(span) | Type::Int(span) | Type::Class(_, span) | Type::Array(_, _, span) => {
                *span
            }
        }
    }
}

impl PartialEq for Type {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Type::Bool(_), Type::Bool(_)) | (Type::Int(_), Type::Int(_)) => true,
            (Type::Array(lhs, _, _), Type::Array(rhs, _, _)) => lhs == rhs,
            _ => false,
        }
    }
}

impl Eq for Type {}

impl fmt::Display for Type {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents an imperative statement.
#[derive(Clone, Debug)]
pub enum Statement {
    /// Assignment
    Assignment(Expression, Expression, Span),

    /// Procedure call
    Call(Call),

    /// Initialization
    Initialization(Initialization),

    /// Variable declaration
    Declaration(Declaration, Span),

    /// Return statement
    Return(Vec<Expression>, Span),

    /// Statement block
    Sequence(Vec<Statement>, Span),

    /// If-else block
    If(Expression, Box<Statement>, Option<Box<Statement>>, Span),

    /// While block
    While(Do, Expression, Box<Statement>, Span),

    /// Break statement
    Break(Span),
}

impl Statement {
    pub fn span(&self) -> Span {
        match self {
            Statement::Call(call) => call.span,
            Statement::Initialization(initialization) => initialization.span,
            Statement::Assignment(_, _, span)
            | Statement::Declaration(_, span)
            | Statement::Return(_, span)
            | Statement::Sequence(_, span)
            | Statement::If(_, _, _, span)
            | Statement::While(_, _, _, span)
            | Statement::Break(span) => *span,
        }
    }
}

impl fmt::Display for Statement {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum Do {
    Yes,
    No,
}

/// Represents an expression (i.e. a term that can be evaluated).
#[derive(Clone, Debug)]
pub enum Expression {
    /// Boolean literal
    Boolean(bool, Span),

    /// Char literal
    Character(char, Span),

    /// String literal
    String(String, Span),

    /// Integer literal
    Integer(i64, Span),

    /// Null literal
    Null(Span),

    /// Class reference
    This(Span),

    /// Variable
    Variable(Symbol, Span),

    /// Array literal
    Array(Vec<Expression>, Span),

    /// Binary operation
    Binary(Cell<Binary>, Box<Expression>, Box<Expression>, Span),

    /// Unary operation
    Unary(Unary, Box<Expression>, Span),

    /// Array index
    Index(Box<Expression>, Box<Expression>, Span),

    /// Array length
    Length(Box<Expression>, Span),

    /// Function call
    Call(Call),

    /// Dot operator
    Dot(Box<Expression>, Symbol, Span),

    /// Class constructor
    New(Symbol, Span),
}

impl Expression {
    pub fn span(&self) -> Span {
        match self {
            Expression::Boolean(_, span)
            | Expression::Character(_, span)
            | Expression::String(_, span)
            | Expression::Integer(_, span)
            | Expression::Null(span)
            | Expression::This(span)
            | Expression::Variable(_, span)
            | Expression::Array(_, span)
            | Expression::Binary(_, _, _, span)
            | Expression::Unary(_, _, span)
            | Expression::Index(_, _, span)
            | Expression::Length(_, span)
            | Expression::Dot(_, _, span)
            | Expression::New(_, span) => *span,
            Expression::Call(call) => call.span,
        }
    }

    pub fn span_mut(&mut self) -> &mut Span {
        match self {
            Expression::Boolean(_, span)
            | Expression::Character(_, span)
            | Expression::String(_, span)
            | Expression::Integer(_, span)
            | Expression::Null(span)
            | Expression::This(span)
            | Expression::Variable(_, span)
            | Expression::Array(_, span)
            | Expression::Binary(_, _, _, span)
            | Expression::Unary(_, _, span)
            | Expression::Index(_, _, span)
            | Expression::Length(_, span)
            | Expression::Dot(_, _, span)
            | Expression::New(_, span) => span,
            Expression::Call(call) => &mut call.span,
        }
    }
}

impl fmt::Display for Expression {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a function call.
#[derive(Clone, Debug)]
pub struct Call {
    pub function: Box<Expression>,
    pub arguments: Vec<Expression>,
    pub span: Span,
}

impl fmt::Display for Call {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum Declaration {
    Multiple(MultipleDeclaration),
    Single(SingleDeclaration),
}

impl Declaration {
    pub fn has_length(&self) -> bool {
        match self {
            Declaration::Multiple(multiple) => multiple.has_length(),
            Declaration::Single(single) => single.has_length(),
        }
    }

    pub fn span(&self) -> Span {
        match self {
            Declaration::Multiple(multiple) => multiple.span(),
            Declaration::Single(single) => single.span(),
        }
    }
}

impl fmt::Display for Declaration {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct MultipleDeclaration {
    pub names: Vec<Symbol>,
    pub r#type: Type,
    pub span: Span,
}

impl MultipleDeclaration {
    pub fn new(names: Vec<Symbol>, r#type: Type, span: Span) -> Self {
        Self {
            names,
            r#type,
            span,
        }
    }

    pub fn has_length(&self) -> bool {
        self.r#type.has_length()
    }

    pub fn span(&self) -> Span {
        self.span
    }
}

impl fmt::Display for MultipleDeclaration {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct SingleDeclaration {
    pub name: Symbol,
    pub r#type: Type,
    pub span: Span,
}

impl SingleDeclaration {
    pub fn new(name: Symbol, r#type: Type, span: Span) -> Self {
        Self { name, r#type, span }
    }

    pub fn has_length(&self) -> bool {
        self.r#type.has_length()
    }

    pub fn span(&self) -> Span {
        self.span
    }
}

impl fmt::Display for SingleDeclaration {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a binary operator.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Binary {
    Mul,
    Hul,
    Div,
    Mod,
    Add,
    Cat,
    Sub,
    Lt,
    Le,
    Ge,
    Gt,
    Eq,
    Ne,
    And,
    Or,
}

impl fmt::Display for Binary {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a unary operator.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Unary {
    Neg,
    Not,
}

impl fmt::Display for Unary {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}
