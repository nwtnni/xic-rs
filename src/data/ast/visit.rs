use crate::data::ast;

pub trait VisitorMut {
    fn visit_interface(&mut self, _interface: &mut ast::Interface) {}

    fn visit_item_signature(&mut self, _item: &mut ast::ItemSignature) {}

    fn visit_class_signature(&mut self, _class: &mut ast::ClassSignature) {}

    fn visit_function_signature(&mut self, _function: &mut ast::FunctionSignature) {}

    fn visit_program(&mut self, _program: &mut ast::Program) {}

    fn visit_use(&mut self, _use: &mut ast::Use) {}

    fn visit_item(&mut self, _item: &mut ast::Item) {}

    fn visit_global(&mut self, _global: &mut ast::Global) {}

    fn visit_class(&mut self, _class: &mut ast::Class) {}

    fn visit_class_item(&mut self, _item: &mut ast::ClassItem) {}

    fn visit_function(&mut self, _function: &mut ast::Function) {}

    fn visit_statement(&mut self, _statement: &mut ast::Statement) {}

    fn visit_declaration(&mut self, _declaration: &mut ast::Declaration) {}

    fn visit_multiple_declaration(&mut self, _declaration: &mut ast::MultipleDeclaration) {}

    fn visit_single_declaration(&mut self, _declaration: &mut ast::SingleDeclaration) {}

    fn visit_initialization(&mut self, _initialization: &mut ast::Initialization) {}

    fn visit_type(&mut self, _type: &mut ast::Type) {}

    fn visit_expression(&mut self, _expression: &mut ast::Expression) {}

    fn visit_call(&mut self, _call: &mut ast::Call) {}

    fn visit_variable(&mut self, _variable: &mut ast::Variable) {}

    fn visit_identifier(&mut self, _identifier: &mut ast::Identifier) {}
}

impl ast::Interface {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::Interface { uses, items } = self;
        uses.iter_mut().for_each(|r#use| r#use.accept_mut(visitor));
        items.iter_mut().for_each(|item| item.accept_mut(visitor));

        visitor.visit_interface(self);
    }
}

impl ast::ItemSignature {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        match self {
            ast::ItemSignature::Class(class) => class.accept_mut(visitor),
            ast::ItemSignature::ClassTemplate(_) => (),
            ast::ItemSignature::Function(function) => function.accept_mut(visitor),
            ast::ItemSignature::FunctionTemplate(_) => (),
        }

        visitor.visit_item_signature(self);
    }
}

impl ast::ClassSignature {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::ClassSignature {
            name,
            extends,
            methods,
            span: _,
        } = self;
        name.accept_mut(visitor);
        extends
            .iter_mut()
            .for_each(|supertype| supertype.accept_mut(visitor));
        methods
            .iter_mut()
            .for_each(|method| method.accept_mut(visitor));

        visitor.visit_class_signature(self);
    }
}

impl ast::FunctionSignature {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::FunctionSignature {
            name,
            parameters,
            returns,
            span: _,
        } = self;
        name.accept_mut(visitor);
        parameters
            .iter_mut()
            .for_each(|parameter| parameter.accept_mut(visitor));
        returns
            .iter_mut()
            .for_each(|r#return| r#return.accept_mut(visitor));

        visitor.visit_function_signature(self);
    }
}

impl ast::Program {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::Program { uses, items } = self;
        uses.iter_mut().for_each(|r#use| r#use.accept_mut(visitor));
        items.iter_mut().for_each(|item| item.accept_mut(visitor));

        visitor.visit_program(self);
    }
}

impl ast::Use {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::Use { name, span: _ } = self;
        visitor.visit_identifier(name);

        visitor.visit_use(self);
    }
}

impl ast::Item {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        match self {
            ast::Item::Global(global) => global.accept_mut(visitor),
            ast::Item::Class(class) => class.accept_mut(visitor),
            ast::Item::ClassTemplate(_) => (),
            ast::Item::Function(function) => function.accept_mut(visitor),
            ast::Item::FunctionTemplate(_) => (),
        }

        visitor.visit_item(self);
    }
}

impl ast::Global {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        match self {
            ast::Global::Declaration(declaration) => declaration.accept_mut(visitor),
            ast::Global::Initialization(initialization) => initialization.accept_mut(visitor),
        }

        visitor.visit_global(self);
    }
}

impl ast::Class {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::Class {
            name,
            extends,
            items,
            span: _,
        } = self;
        name.accept_mut(visitor);
        extends
            .iter_mut()
            .for_each(|supertype| supertype.accept_mut(visitor));
        items.iter_mut().for_each(|item| item.accept_mut(visitor));

        visitor.visit_class(self);
    }
}

impl ast::ClassItem {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        match self {
            ast::ClassItem::Field(field) => field.accept_mut(visitor),
            ast::ClassItem::Method(method) => method.accept_mut(visitor),
        }

        visitor.visit_class_item(self);
    }
}

impl ast::Function {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::Function {
            name,
            parameters,
            returns,
            statements,
            span: _,
        } = self;
        name.accept_mut(visitor);
        parameters
            .iter_mut()
            .for_each(|parameter| parameter.accept_mut(visitor));
        returns
            .iter_mut()
            .for_each(|r#return| r#return.accept_mut(visitor));
        statements.accept_mut(visitor);

        visitor.visit_function(self);
    }
}

impl ast::Statement {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        match self {
            ast::Statement::Assignment(destination, source, _) => {
                destination.accept_mut(visitor);
                source.accept_mut(visitor);
            }
            ast::Statement::Call(call) => call.accept_mut(visitor),
            ast::Statement::Initialization(initialization) => initialization.accept_mut(visitor),
            ast::Statement::Declaration(declaration, _) => declaration.accept_mut(visitor),
            ast::Statement::Return(r#returns, _) => r#returns
                .iter_mut()
                .for_each(|r#return| r#return.accept_mut(visitor)),
            ast::Statement::Sequence(statements, _) => statements
                .iter_mut()
                .for_each(|statement| statement.accept_mut(visitor)),
            ast::Statement::If(condition, r#if, r#else, _) => {
                condition.accept_mut(visitor);
                r#if.accept_mut(visitor);
                if let Some(r#else) = r#else {
                    r#else.accept_mut(visitor);
                }
            }
            ast::Statement::While(_, condition, r#while, _) => {
                condition.accept_mut(visitor);
                r#while.accept_mut(visitor);
            }
            ast::Statement::Break(_) => (),
        }

        visitor.visit_statement(self);
    }
}

impl ast::Declaration {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        match self {
            ast::Declaration::Multiple(declaration) => declaration.accept_mut(visitor),
            ast::Declaration::Single(declaration) => declaration.accept_mut(visitor),
        }

        visitor.visit_declaration(self);
    }
}

impl ast::MultipleDeclaration {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::MultipleDeclaration {
            names,
            r#type,
            span: _,
        } = self;
        names.iter_mut().for_each(|name| name.accept_mut(visitor));
        r#type.accept_mut(visitor);

        visitor.visit_multiple_declaration(self);
    }
}

impl ast::SingleDeclaration {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::SingleDeclaration {
            name,
            r#type,
            span: _,
        } = self;
        name.accept_mut(visitor);
        r#type.accept_mut(visitor);

        visitor.visit_single_declaration(self);
    }
}

impl ast::Initialization {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::Initialization {
            declarations,
            expression,
            span: _,
        } = self;
        declarations
            .iter_mut()
            .flatten()
            .for_each(|declaration| declaration.accept_mut(visitor));
        expression.accept_mut(visitor);

        visitor.visit_initialization(self);
    }
}

impl ast::Type {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        match self {
            ast::Type::Bool(_) => (),
            ast::Type::Int(_) => (),
            ast::Type::Class(variable) => variable.accept_mut(visitor),
            ast::Type::Array(r#type, length, _) => {
                r#type.accept_mut(visitor);
                if let Some(length) = length {
                    length.accept_mut(visitor);
                }
            }
        }

        visitor.visit_type(self);
    }
}

impl ast::Expression {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        match self {
            ast::Expression::Boolean(_, _)
            | ast::Expression::Character(_, _)
            | ast::Expression::String(_, _)
            | ast::Expression::Integer(_, _)
            | ast::Expression::Null(_)
            | ast::Expression::This(_)
            | ast::Expression::Super(_) => (),
            ast::Expression::Variable(variable) => variable.accept_mut(visitor),
            ast::Expression::Array(expressions, _) => expressions
                .iter_mut()
                .for_each(|expression| expression.accept_mut(visitor)),
            ast::Expression::Binary(_, left, right, _) => {
                left.accept_mut(visitor);
                right.accept_mut(visitor);
            }
            ast::Expression::Unary(_, expression, _) => expression.accept_mut(visitor),
            ast::Expression::Index(array, index, _) => {
                array.accept_mut(visitor);
                index.accept_mut(visitor);
            }
            ast::Expression::Length(array, _) => {
                array.accept_mut(visitor);
            }
            ast::Expression::Call(call) => call.accept_mut(visitor),
            ast::Expression::Dot(_, receiver, identifier, _) => {
                receiver.accept_mut(visitor);
                identifier.accept_mut(visitor);
            }
            ast::Expression::New(variable, _) => variable.accept_mut(visitor),
        }

        visitor.visit_expression(self);
    }
}

impl ast::Call {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::Call {
            function,
            arguments,
            span: _,
        } = self;
        function.accept_mut(visitor);
        arguments
            .iter_mut()
            .for_each(|argument| argument.accept_mut(visitor));

        visitor.visit_call(self);
    }
}

impl ast::Variable {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        let ast::Variable {
            name,
            generics,
            span: _,
        } = self;
        name.accept_mut(visitor);
        generics
            .iter_mut()
            .flatten()
            .for_each(|r#type| r#type.accept_mut(visitor));

        visitor.visit_variable(self);
    }
}

impl ast::Identifier {
    pub fn accept_mut<V: VisitorMut>(&mut self, visitor: &mut V) {
        visitor.visit_identifier(self);
    }
}
