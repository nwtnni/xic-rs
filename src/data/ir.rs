use std::collections::HashMap;

use crate::data::hir;
use crate::data::lir;
use crate::data::operand;
use crate::util::symbol;

#[derive(Clone, Debug)]
pub struct Unit<F: IR> {
    pub name: symbol::Symbol,
    pub funs: HashMap<symbol::Symbol, F>,
    pub data: HashMap<symbol::Symbol, operand::Label>,
}

pub trait IR {}
impl IR for hir::Fun {}
impl IR for lir::Fun {}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Bin {
    Add,
    Sub,
    Mul,
    Hul,
    Div,
    Mod,
    Xor,
    Ls,
    Rs,
    ARs,
    Lt,
    Le,
    Ge,
    Gt,
    Ne,
    Eq,
    And,
    Or,
}
