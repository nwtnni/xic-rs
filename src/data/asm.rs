use crate::data::ir;
use crate::data::operand;

pub type Unit<T> = ir::Unit<Function<T>>;
pub type Function<T> = ir::Function<Vec<Assembly<T>>>;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Assembly<T> {
    Binary(Binary, operand::Binary<T>),
    Unary(Unary, operand::Unary<T>),
    Nullary(Nullary),
    Label(operand::Label),
    Directive(Directive),
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Directive {
    Intel,
    Align(usize),
    Local(operand::Label),
    Global(operand::Label),
    Quad(Vec<i64>),
    Data,
    Text,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Binary {
    Add,
    Sub,
    And,
    Or,
    Xor,
    Cmp,
    Mov,
    Lea,
}

impl From<ir::Binary> for Binary {
    fn from(binary: ir::Binary) -> Self {
        match binary {
            ir::Binary::Add => Binary::Add,
            ir::Binary::Sub => Binary::Sub,
            ir::Binary::Xor => Binary::Xor,
            ir::Binary::And => Binary::And,
            ir::Binary::Or => Binary::Or,
            _ => panic!("[INTERNAL ERROR]: converting unsupported IR operator"),
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Unary {
    Neg,
    Push,
    Pop,
    Call { arguments: usize, returns: usize },
    Mul,
    Div(Division),
    Jmp,
    Jcc(Condition),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Nullary {
    Cqo,
    Ret,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Condition {
    L,
    Le,
    Ge,
    G,
    Ne,
    E,
}

impl From<ir::Condition> for Condition {
    fn from(condition: ir::Condition) -> Self {
        match condition {
            ir::Condition::Lt => Condition::L,
            ir::Condition::Le => Condition::Le,
            ir::Condition::Ge => Condition::Ge,
            ir::Condition::Gt => Condition::G,
            ir::Condition::Ne => Condition::Ne,
            ir::Condition::Eq => Condition::E,
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Division {
    Quotient,
    Remainder,
}
