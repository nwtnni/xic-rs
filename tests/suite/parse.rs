use std::fmt;
use std::path::Path;

use xic::data::ast;

struct Snapshot(Result<ast::Program, xic::Error>);

impl fmt::Display for Snapshot {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        match &self.0 {
            Ok(program) => write!(fmt, "{}", program),
            Err(error) => match error.report() {
                None => write!(fmt, "{}", error),
                Some(report) => {
                    let cache = xic::data::span::FileCache::default();
                    let mut buffer = Vec::new();

                    report
                        .with_config(ariadne::Config::default().with_color(false))
                        .finish()
                        .write(cache, &mut buffer)
                        .map_err(|_| fmt::Error)?;

                    write!(fmt, "{}", String::from_utf8(buffer).unwrap())
                }
            },
        }
    }
}

#[test_generator::test_resources("tests/parse/*.xi")]
pub fn parse(path: &str) {
    let tokens = xic::api::lex(Path::new(path)).unwrap();
    let program = xic::api::parse(tokens);

    insta::assert_display_snapshot!(path, Snapshot(program));
}
