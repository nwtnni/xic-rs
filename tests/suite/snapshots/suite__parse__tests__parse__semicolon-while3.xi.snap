---
source: tests/suite/parse.rs
assertion_line: 22
expression: Snapshot(program)
---
(() ((foo () () ((while false ((= (x int) 5) (= x (bar)) (foo2) (= x 5)))))))
